src/common/dom/apply_themes_on_element.ts

```diff
@@ -23,57 +23,104 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
+    // Determine the primary and accent colors (fallbacks are the HA default blue and orange).
+    // Currently only the default theme offers color pickers, so for all other themes
+    // we have to check if the theme explicitly defines a color value for them.
+    let primaryColor;
+    let accentColor;
+    if (selectedTheme === "default") {
+      // User selected colors
+      primaryColor = themeSettings.primaryColor;
+      accentColor = themeSettings.accentColor;
+    } else if (selectedTheme && themes.themes[selectedTheme]) {
+      // Values from theme or default style
+      if (themeSettings.dark) {
+        primaryColor = themes.themes[selectedTheme].dark_styles
+          ? themes.themes[selectedTheme].dark_styles!["primary-color"]
+          : darkStyles["primary-color"];
+        accentColor = themes.themes[selectedTheme].dark_styles
+          ? themes.themes[selectedTheme].dark_styles!["accent-color"]
+          : darkStyles["accent-color"];
+      } else {
+        const newScheme = themes.themes[selectedTheme].styles !== undefined;
+        primaryColor = newScheme
+          ? themes.themes[selectedTheme].styles!["primary-color"]
+          : themes.themes[selectedTheme]["primary-color"];
+        accentColor = newScheme
+          ? themes.themes[selectedTheme].styles!["accent-color"]
+          : themes.themes[selectedTheme]["accent-color"];
+      }
+    }
+    primaryColor = primaryColor || "#03a9f4";
+    accentColor = accentColor || "#ff9800";
+
+    if (themeSettings.dark) {
       cacheKey = `${cacheKey}__dark`;
       themeRules = darkStyles;
-      if (themeOptions.primaryColor) {
-        themeRules["app-header-background-color"] = hexBlend(
-          themeOptions.primaryColor,
-          "#121212",
-          8
-        );
-      }
+
+      themeRules["app-header-background-color"] = hexBlend(
+        primaryColor,
+        "#121212",
+        8
+      );
     }
-    if (themeOptions.primaryColor) {
-      cacheKey = `${cacheKey}__primary_${themeOptions.primaryColor}`;
-      const rgbPrimaryColor = hex2rgb(themeOptions.primaryColor);
+
+    if (primaryColor) {
+      cacheKey = `${cacheKey}__primary_${primaryColor}`;
+      const rgbPrimaryColor = hex2rgb(primaryColor);
       const labPrimaryColor = rgb2lab(rgbPrimaryColor);
-      themeRules["primary-color"] = themeOptions.primaryColor;
-      const rgbLigthPrimaryColor = lab2rgb(labBrighten(labPrimaryColor));
-      themeRules["light-primary-color"] = rgb2hex(rgbLigthPrimaryColor);
+      themeRules["primary-color"] = primaryColor;
+      const rgbLightPrimaryColor = lab2rgb(labBrighten(labPrimaryColor));
+      themeRules["light-primary-color"] = rgb2hex(rgbLightPrimaryColor);
       themeRules["dark-primary-color"] = lab2hex(labDarken(labPrimaryColor));
       themeRules["text-primary-color"] =
-        rgbContrast(rgbPrimaryColor, [33, 33, 33]) < 6 ? "#fff" : "#212121";
+        rgbContrast(rgbPrimaryColor, [33, 33, 33]) < 7 ? "#fff" : "#212121";
```

Comment:
Previously we only executed this calculation if we had a user overridden primary color. That meant that if the user did not change it, we would stay with the default white on blue color for the header (since that is using `text-primary-color`). However those default color values do not follow the contrast value 6. To keep the look consistent with how it was before for the default theme I adjusted to 7. That however could mean that for users that manually specified a color for the default theme right at the cutoff point between contrast 6 and 7, their header color might change.

Alternative is to add a guard around  `text-primary-color` and related keys again to only execute it if the user overrides the colors in the picker. However, even only moving the default color by a single RGB value (smallest possible change), means that then this contrast logic takes effect and switches from white on blue to black on blue in the header, so we would only be hiding the fact that our own default colors do not conform with our defined desired contrast value.

Reply:
Reverted since color picker support was removed and so this contrast issue is now less likely to be noticed (although I still think the old logic is flawed).